[Metrics] [Docs] Add section about device placement (#5280)
PyTorchLightning/pytorch-lightning: docs/source/metrics.rst

    To change this, after initializing the metric, the method ``.persistent(mode)`` can
    be used to enable (``mode=True``) or disable (``mode=False``) this behaviour.

*******************
Metrics and devices
*******************

> Metrics are simple subclasses of ``torch.nn.Module`` and their metric states behave

Reviewer: ```suggestion
Metrics are simple subclasses of :class:`~torch.nn.Module` and their metric states behave
```

similar to buffers and parameters of modules. This means that metrics states should
be moved to the same device as the input of the metric:

.. code-block:: python

    import torch
    from pytorch_lightning.metrics import Accuracy

    target = torch.tensor([1, 1, 0, 0], device=torch.device("cuda", 0))
    preds = torch.tensor([0, 1, 0, 0], device=torch.device("cuda", 0))

    # Metric states are always initialized on cpu, and needs to be moved to
    # the correct device
    confmat = Accuracy(num_classes=2).to(torch.device("cuda", 0))
    out = confmat(preds, target)
    print(out.device) # cuda:0

> However, when **proper defined** inside an ``LightningModule``, lightning will

Reviewer 2: ```suggestion
However, when **properly defined** inside an ``LightningModule``, lightning will
```

> However, when **properly defined** inside a ``LightningModule``, lightning will

Reviewer 3: ```suggestion
However, when **properly defined** inside a ``LightningModule``, Lightning will
```
I think generally the Lightning brand is capitalized.

> However, when **properly defined** inside an ``LightningModule``, lightning will

Reviewer 4: ```suggestion
However, when **properly defined** inside a ``LightningModule``, lightning will
```

> automatically move the metrics to the same device as the data. By **proper defined**

Reviewer 2: ```suggestion
automatically move the metrics to the same device as the data. By **properly defined**
```

> automatically move the metrics to the same device as the data. By **properly defined**

Reviewer 4: ```suggestion
automatically move the metrics to the same device as the data. Being **properly defined**
```

> it is meant that the metric is correctly identified as a child module of the model 

Reviewer 4: ```suggestion
means that the metric is correctly identified as a child module of the model 
```

(check ``.children()`` attribute of the model). Therefore, metrics cannot be placed 
in native python ``list`` and ``dict``, as they will not be correctly identified

> as child modules. Instead of ``list`` use ``torch.nn.ModuleList`` and instead of

Reviewer 3: ```suggestion
as child modules. Instead of ``list`` use :class:`torch.nn.ModuleList` and instead of
```
btw we can directly link torch docs. Hope I didn't make a typo

Reviewer: just missed `~`

Reviewer 3: it's optional, it depends if you want to show short or long name :)

> as a child modules. Instead of ``list`` use ``torch.nn.ModuleList`` and instead of

Reviewer 2: ```suggestion
as child modules. Instead of ``list`` use ``torch.nn.ModuleList`` and instead of
```

> ``dict`` use ``torch.nn.ModuleDict``.

Reviewer 3: ```suggestion
``dict`` use :class:`torch.nn.ModuleDict`.
```

> .. code-block::

Reviewer 5: can this be test too?
```suggestion
.. testcode::
```

    To change this, after initializing the metric, the method ``.persistent(mode)`` can
    be used to enable (``mode=True``) or disable (``mode=False``) this behaviour.

*******************
Metrics and devices
*******************

Metrics are simple subclasses of :class:`~torch.nn.Module` and their metric states behave
similar to buffers and parameters of modules. This means that metrics states should
be moved to the same device as the input of the metric:

> .. code-block:: python

Reviewer: ```suggestion
.. testcode::
```
I hope it's correct.

    import torch
    from pytorch_lightning.metrics import Accuracy

    target = torch.tensor([1, 1, 0, 0], device=torch.device("cuda", 0))
    preds = torch.tensor([0, 1, 0, 0], device=torch.device("cuda", 0))

    # Metric states are always initialized on cpu, and needs to be moved to
    # the correct device
    confmat = Accuracy(num_classes=2).to(torch.device("cuda", 0))
    out = confmat(preds, target)
    print(out.device) # cuda:0

However, when **properly defined** inside a :class:`~pytorch_lightning.core.lightning.LightningModule`

> , Lightning will
> automatically move the metrics to the same device as the data. Being **properly defined**
> means that the metric is correctly identified as a child module of the model 
> (check ``.children()`` attribute of the model). Therefore, metrics cannot be placed 

Reviewer: ```suggestion
, Lightning will automatically move the metrics to the same device as the data. Being
**properly defined** means that the metric is correctly identified as a child module of the
model (check ``.children()`` attribute of the model). Therefore, metrics cannot be placed
```
